src/vsr/replica.zig

```diff
@@ -2642,6 +3054,65 @@ pub fn Replica(
             return false;
         }
 
+        /// Returns whether the highest known op is certain.
+        ///
+        /// After recovering the WAL, there are 2 possible outcomes:
+        /// * All entries valid. The highest op is certain, and safe to set as `replica.op`.
+        /// * One or more entries are faulty. The highest op isn't certain — it may be one of the
+        ///   broken entries.
+        ///
+        /// The replica must refrain from repairing any faulty slots until the highest op is known.
+        /// Otherwise, if we were to repair a slot while uncertain of `replica.op`:
+        ///
+        /// * we may nack an op that we shouldn't, or
+        /// * we may replace a prepared op that we were guaranteeing for the primary, potentially
+        ///   forking the log.
+        ///
+        ///
+        /// Test for a fault the right of the current op. The fault might be our true op, and
+        /// sharing our current `replica.op` might cause the cluster's op to likewise regress.
+        ///
+        /// Note that for our purposes here, we only care about entries that were faulty during
+        /// WAL recovery, not ones that were found to be faulty after the fact (e.g. due to
+        /// `request_prepare`).
+        ///
+        /// Cases (`✓`: `replica.op_checkpoint`, `✗`: faulty, `o`: `replica.op`):
+        /// * ` ✓ o ✗ `: View change is unsafe.
+        /// * ` ✗ ✓ o `: View change is unsafe.
+        /// * ` ✓ ✗ o `: View change is safe.
+        /// * ✓=o:       View change is unsafe if any slots are faulty
+        ///              (`replica.op_checkpoint` == `replica.op`).
+        // TODO Use this function once we switch from recovery protocol to the superblock.
+        // If there is an "unsafe" fault, we will need to request a start_view from the leader to
+        // learn the op.
+        fn op_certain(self: *const Self) bool {
+            assert(self.status == .recovering);
+            assert(self.journal.recovered);
+            assert(self.op_checkpoint <= self.op);
+
+            const slot_op_checkpoint = self.journal.slot_for_op(self.op_checkpoint);
+            const slot_op = self.journal.slot_with_op(self.op).?;
+            const slot_op_min = std.math.min(slot_op, slot_op_checkpoint);
+            const slot_op_max = std.math.max(slot_op, slot_op_checkpoint);
```

Comment:
I didn't even realize that. It also (`<=` and `>=`) `Slot`s directly — I didn't realize that worked. What are the rules for this in Zig?

Reply:
```zig
test "min implicitly works on non-integers" {
    const std = @import("std");
    const expectEqual = std.testing.expectEqual;
    
    const Slot = struct { index: u64 };

    const b = Slot{ .index = 3 };
    const a = Slot{ .index = 2 };

    expectEqual(a, std.math.min(b, a));
}
```

It actually doesn't work:

```bash
./tigerbeetle/zig/lib/std/math.zig:334:27: error: expected type 'Slot', found 'comptime_int'
    return @TypeOf(@as(A, 0) + @as(B, 0));
```

I think the reason this code appears to compile, is because it doesn't—we don't call `op_certain()` from anywhere yet.